Refactor the standard integration guide

gregjopa opened the pull request:
The DevRel team has been working on improving the Standard Integration code snippets across developer.paypal.com. This PR includes many of these new changes.

Improvements include:
- Using the common naming of PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET for env variables.
- Using a single file for the Node.js server code.
- Teach negative testing with code comments for using the `PayPal-Mock-Response` request header for order creation and capture.
- Always try to relay PayPal API JSON responses and status codes back to the frontend.
- Death to `alert()` since it pauses the popup flow. Replacing this with appending UI messages to the button container.

There are some other changes I am less confident on and could use feedback on.
- What route names should we use for wrapping create order and capture order api calls?
- Whats the best way to pass the OrderID to the create order api wrapper? Pass it in the URL or in the post body?
- Should we recommend the Express static module or just return the html file directly from the file system?

Feedback welcome! Let us know what you like and don't like.

wsbrunson commented on standard-integration/index.html:
```diff
@@ -0,0 +1,119 @@
+  <body>
+    <div id="paypal-button-container"></div>
+    <!-- Replace the "test" client-id value with your client-id -->
+    <script src="https://www.paypal.com/sdk/js?client-id=test&currency=USD"></script>
+    <script>
```
one thing i would love to see is all of our examples using `paypal-js`

gregjopa replied:
I'd be open to that. How do you think they should use `paypal-js` in a repo like this? Should we recommend a separate client-side js file that imports in paypal-js and then serve up a bundle via rollup or webpack? Or host it from paypalobjects.com and have folks download it from there?

wsbrunson replied:
I think the later. I think ideally we'd have something that was as easy as stripe's: https://stripe.com/docs/libraries/stripejs-esmodule#web-stripejs-html

wsbrunson replied:
I think with the former, we have to start talking about build tools and that always sucks

gregjopa replied:
I hear ya on the burden of teaching build tools. I like offering both the CDN option and the ES Module like Stripe does. I'll take this paypal-js loader suggestion back to the team to get more feedback.

wsbrunson reviewed (approved):
looks amazing!

jshawl commented on standard-integration/server.js:
```diff
@@ -1,34 +1,149 @@
 // parse post params sent in body in json format
 app.use(express.json());
 
-app.post("/my-server/create-paypal-order", async (req, res) => {
```
💯 this tripped me up the other day thinking `my-server` was not an actual route...

jshawl commented on standard-integration/server.js:
```diff
@@ -1,34 +1,149 @@
+    res.status(httpStatusCode).json(jsonResponse);
+  } catch (error) {
+    console.error('Failed to create order:', error);
+    res.status(500).json({ error: 'Failed to create order.' });
```
will the server 500 with the stack trace without this `catch`?

MALIKSHAHID786 replied:
Hi

MALIKSHAHID786 replied:
I am happy

jshawl commented on standard-integration/index.html:
```diff
@@ -0,0 +1,119 @@
+                    {
+                      id: 'YOUR_PRODUCT_ID',
+                      quantity: 'YOUR_PRODUCT_QUANTITY',
+                    },
```
option to include `purchase_units` as an example here

gregjopa replied:
Unfortunately we cannot recommend having purchase_units in the frontend payload, at least not in the Orders API format. The amount has to be managed on the server-side to prevent any malicious actors from tampering with it via Chrome Dev Tools.

jshawl commented on standard-integration/server.js:
```diff
@@ -1,34 +1,149 @@
+});
+
+app.listen(8888, () => {
+  console.log('listening on http://localhost:8888/');
```
option to use a `PORT` env var here that defaults to 8888

gregjopa replied:
Great suggestion @jshawl! I just pushed a commit to optionally read PORT from an env and default to 8888

jshawl commented on standard-integration/server.js:
```diff
@@ -1,34 +1,149 @@
-import * as paypal from "./paypal-api.js";
-const {PORT = 8888} = process.env;
+import express from 'express';
+import fetch from 'node-fetch';
```
option to add a `.nvmrc` in this directory with `14` to match [the readme](https://github.com/paypal-examples/docs-examples#have-nodejs-installed)

gregjopa replied:
Oh wow Node 14 is no longer even supported. Good call out. I updated the doc to recommend Node 16. We can revisit adding an .nvmrc file and/or validate the minimum required Node.js version on server startup in another PR.

jshawl commented on standard-integration/server.js:
```diff
@@ -1,34 +1,149 @@
+      // https://developer.paypal.com/tools/sandbox/negative-testing/request-headers/
+      // "PayPal-Mock-Response": '{"mock_application_codes": "MISSING_REQUIRED_PARAMETER"}'
+      // "PayPal-Mock-Response": '{"mock_application_codes": "PERMISSION_DENIED"}'
+      // "PayPal-Mock-Response": '{"mock_application_codes": "INTERNAL_SERVER_ERROR"}'
```
👏 👏 👏

westeezy replied:
These are great! I think it'd also be nice to add to a section to https://developer.paypal.com/sdk/js/ with something like `Troubleshooting` and these would show: "negative testing is done via the following headers"

gregjopa replied:
Great point @westeezy. Let's figure out how to best add content about these headers to the sdk docs.

jshawl reviewed (approved):
Looks great! Left a few non-blocking comments :shipit:

MALIKSHAHID786 commented:
Ravievd payment

gregjopa commented:
Thanks for the feedback everyone! I applied some of the recommendations and I'm going to merge this PR.

@cnallam, fyi I changed two things with the codespaces integration for standard checkout:
1. There was a bug with the workspace path that is now updated.
2. The env variables have been renamed to PAYPAL_CLIENT_ID and PAYPAL_CLIENT_SECRET to better match the PayPal Auth docs and to help the developer know that these credentials are for PayPal. We plan to make the same change to the Advanced Integration in a follow up PR.

I've successfully tested the CodeSpaces integration off this feature branch. Here are some screenshots for reference:

<img width="1036" alt="Screenshot 2023-08-15 at 9 48 08 AM" src="https://github.com/paypal-examples/docs-examples/assets/534034/96d02ae1-2f50-40e5-8b61-7dfb90d8e0da">
<img width="1275" alt="Screenshot 2023-08-15 at 9 51 14 AM" src="https://github.com/paypal-examples/docs-examples/assets/534034/46656068-8125-40cd-b794-5b24c4b524db">
<img width="1094" alt="Screenshot 2023-08-15 at 9 51 21 AM" src="https://github.com/paypal-examples/docs-examples/assets/534034/e787f289-3e5c-41be-a34b-64fc17d3070e">
